UnitControl: Add support for unit step per unit type

rmorse opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
The [UnitControl](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/unit-control) component does not support a "step" value.  Therefor it always increments exactly by 1.

This is probably not desired for certain unit types (i.e., rem, em, etc) in certain situations.

This update, allows for adding a "step" property to the units argument - so it can be controlled on a per unit type basis:

So we can now do this:
```
    const units = [
    	{ value: 'px', label: 'px', default: 0, step: 1 },
    	{ value: '%', label: '%', default: 10, step: 1 },
    	{ value: 'em', label: 'em', default: 0, step: 0.1 },
    ];
```
It defaults to a value of 1, if the unit type does not contain a step (for backwards compatibility).

Personally, I wanted to use this as an improved "Line height" control, that also supports units, and the first unit is blank:
`{ value: '', label: '', default: 1, step: 0.1 },` 
But I think there is a good argument for supporting different steps for different use cases on a per unit basis.


## How has this been tested?
I've tested the component by following the example, by suppling a step, and without.

All incremements behave as expected.

**I have not tested the React Native stuff** - but I have updated the code.  I checked, and it looks like `RangeCell` and `StepperCell` both have a `step` property so this should work - can someone test this? (I'm not sure how to)

## Screenshots <!-- if applicable -->
![unitcontrol-step](https://user-images.githubusercontent.com/853547/112971917-fee81a80-914f-11eb-97d4-4955a0ff6322.gif)
Don't worry about the value reset on unit change - this is not hooked up to state.
This example uses the following props for unit types (notice 0.1 step on `em`, `rem`, and *empty*)
```
const unitTypes = [
	{ value: '', label: '', default: 1, step: 0.1 },
	{ value: 'px', label: 'px', default: 0 },
	{ value: '%', label: '%', default: 0 },
	{ value: 'em', label: 'em', default: 0, step: 0.1 },
	{ value: 'rem', label: 'rem', default: 0, step: 0.1 },
	{ value: 'vw', label: 'vw', default: 0 },
	{ value: 'vh', label: 'vh', default: 0 },
];
```

## Types of changes
Minor improvement - we check the unit types argument, for a `step` value and pass that into our number input, if it is not present, default to a step of `1`
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

*Update - I guess this also builds on: https://github.com/WordPress/gutenberg/pull/31314

rmorse committed:
add support for unit step per unit type

rmorse committed:
remove default step value from CSS_UNITS const

rmorse committed:
add react native support

rmorse commented on packages/components/src/unit-control/index.js:
```diff
@@ -22,6 +22,7 @@ import { Root, ValueInput } from './styles/unit-control-styles';
 import UnitSelectControl from './unit-select-control';
 import { CSS_UNITS, getParsedValue, getValidParsedUnit } from './utils';
 import { useControlledState } from '../utils/hooks';
+// import { active } from '@wp-g2/styles/types';
```
This was causing my local build to fail, and it wasn't in use - so I've commented it for now

mcsf replied:
Do we still need anything here?

rmorse committed:
missed step on native RangeCell
add comments

rmorse commented:
@ItsJonQ - thought I would flag you on this as you've been working on the previous iterations and will have a sense if this is in the right direction.

paaljoachim commented:
@aristath @gziolo @mcsf Can you take a look at this PR?
Thanks!

aristath commented:
What are all the changes to `package-lock.json` for?

rmorse commented:
Hmm there shouldn't have been, maybe a bad merge on my side, I'll double check later on today and see if I can get this tidied up.

mcsf commented:
@rmorse: thanks for the PR. Ensuring that certain units, like `em`, can default to stepping by `0.1` makes sense as far as the user is concerned. However, I wanted to correct the opening assumption:

> The [UnitControl](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/unit-control) component does not support a "step" value. Therefor it always increments exactly by 1.

The component does accept `step` as a prop, which it will pass down until it reaches `NumberControl`. I agree that this isn't clear, and the inheritance model used by the `*Control` components makes it hard to determine what the full API is.

@nosolosw: is the addition of units to `line-height` compatible with what you're doing in Global Styles?

If so, then @rmorse, we could go with your suggestion to provide `step` defaults for each unit in `CSS_UNITS`, as long as we don't override any incoming `step` prop:

```js
let step = stepProp;
if ( ! step ) {
  const activeUnit = units.find( ( option ) => option.value === unit );
  step = activeUnit?.step ? activeUnit.step : 1;
}
```

My last note is to treat the non-unit, `''`, locally in LineHeightControl and not in UnitControl. That's because, in other contexts, it may refer to larger or strictly integer values (e.g. `columns: 2`).

oandregal commented:
Things that relate to this from the "theme.json" perspective are:

- There's interest in adding units other than pixels for font sizes https://github.com/WordPress/gutenberg/issues/25137 https://github.com/WordPress/gutenberg/issues/23323 so I wonder how/whether these are related to each other.

- We do have some `units` in the theme.json spec https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#settings although they seem to affect only spacing properties (margin, padding), as far as I know. If we add these in the line-height control and others, we may want to add the corresponding setting in theme.json for themes to control which ones to display.

rmorse committed:
resolve package.json conflict, update repo

rmorse committed:
allow props to override unit steps

rmorse commented:
@mcsf - I've updated that.  To confirm logic:

- The UnitControl can accept a step prop, and that should supercede the step value associated with units
- It is expected that this prop, is applied to all units when passed

I've added your suggestion to the PR (and also added a condition to default to the first unit type, if somehow one wasn't found)

mcsf commented on packages/components/src/unit-control/index.js:
```diff
@@ -159,11 +159,18 @@ function UnitControl(
+	if ( ! step ) {
+		// Get current unit type
+		let activeUnit = units.find( ( option ) => option.value === unit );
+		// if unit wasn't found, choose the first option
+		if ( ! activeUnit && units.length > 0 ) {
+			activeUnit = units[ 0 ];
+		}
```
If we can't find a matching unit, it doesn't really make sense to grab the first one. We should instead leave early with `step = 1`.

rmorse replied:
I'm inclined to agree, however to explain my logic - when I was doing some testing (putting in an empty value), the unit was not detected (no active unit).

In this scenario, it looks like the first unit (from the list of units) is selected (visually) due to the nature of a select even when it has no value.  This just meant, even in those edge cases, we were still respecting the step associated with the unit type - but I guess that is more or less catering for a unlikely/unwanted use case?

rmorse replied:
Removed this in the latest commit.

mcsf commented on packages/components/src/unit-control/index.js:
```diff
@@ -159,11 +159,18 @@ function UnitControl(
+			activeUnit = units[ 0 ];
+		}
+		// Extract the step if it exists, or default to `1`
+		step = activeUnit?.step ? activeUnit.step : 1;
```
Now with the [nullish coalescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator) we can use:

```suggestion
		step = activeUnit?.step ?? 1;
```

rmorse replied:
Awesome, haven't got these on reflex yet ^^

mcsf commented on packages/components/src/unit-control/index.js:
```diff
@@ -158,6 +159,19 @@ function UnitControl(
 	) : null;
 
+	let step = props.step;
+	// if no step prop has been passed
```
I would replace these micro-level comments with a brief high-level comment, e.g.

> Unless a `step` value is explicitly passed, look for any unit-specific values. For instance, a sensible step value for `em` might be `0.1` instead of `1`.

rmorse replied:
Thanks for the feedback, I'll look to update all this again tomorrow.

rmorse replied:
updated, thanks

mcsf reviewed (commented):
Thanks for cleaning up. Here's my review.

rmorse committed:
Update packages/components/src/unit-control/index.js

Co-authored-by: Miguel Fonseca <miguelcsf@gmail.com>

rmorse committed:
updates from review ->
remove setting default first unit
tidy code + comments

rmorse committed:
Merge branch 'trunk' into update/unitcontrol

rmorse committed:
Use capitals in comments

rmorse committed:
Use block commments instead of single line

mcsf commented on packages/components/src/unit-control/index.native.js:
```diff
@@ -149,6 +161,7 @@ function UnitControl( {
 					minimumValue={ min }
 					maximumValue={ max }
 					value={ value }
+					step={ unitStep }
```
Is this intentional? Or should it be:

```suggestion
					step={ step }
```

mcsf commented on packages/components/src/unit-control/index.js:
```diff
@@ -158,6 +158,17 @@ function UnitControl(
+	 * If no step prop has been passed, lookup the active unit and
+	 * try to get step from `units`, or default to a value of `1`
+	 */
+	if ( ! step ) {
```
Per the failing unit tests, the `units` prop may be `false`, so we need:

```suggestion
	if ( ! step && units ) {
```

mcsf commented on packages/components/src/unit-control/index.native.js:
```diff
@@ -123,6 +123,17 @@ function UnitControl( {
+	 * If no step prop has been passed, lookup the active unit and
+	 * try to get step from `units`, or default to a value of `1`
+	 */
+	if ( ! step ) {
```
Per the failing unit tests, the `units` prop may be `false`, so we need:

```suggestion
	if ( ! step && units ) {
```

mcsf reviewed (changes requested):
Hey, @rmorse, we're almost there! Please apply these last fixes, and we should be good to go.

rmorse committed:
Apply suggestions from code review

Co-authored-by: Miguel Fonseca <miguelcsf@gmail.com>

rmorse committed:
Merge branch 'trunk' into update/unitcontrol

rmorse commented:
Thanks @mcsf , now done - I didn't understand those fails on `units` being able to be `false` - should have spotted the bad variable name though ;)

mcsf reviewed (approved):
🚢

dcalhoun reviewed (approved):
LGTM. I tested the changes for native on an iPhone 11 simulator and Samsung Galaxy S20 to test for any regressions. 

I also opened https://github.com/wordpress-mobile/gutenberg-mobile/pull/3479 to allow additional native CI tests to run for this change. I'd recommend awaiting those to pass before merging.

dcalhoun commented:
The current failures in wordpress-mobile/gutenberg-mobile#3479 are unrelated to the changes in this PR and have been resolved in https://github.com/wordpress-mobile/gutenberg-mobile/pull/3477. We should be good to go for merging this work. 👍🏻

mcsf commented:
Thanks for everyone’s help!
